runtime/near-vm-logic/src/gas_counter.rs

```diff
@@ -49,48 +69,98 @@ impl GasCounter {
-            used_gas: 0,
-            max_gas_burnt,
+            fast_counter: FastGasCounter {
+                burnt_gas: 0,
+                gas_limit: if is_view {
+                    // Ignore prepaid gas limit and promises.
+                    max_gas_burnt
+                } else {
+                    min(max_gas_burnt, prepaid_gas)
+                },
+                opcode_cost: Gas::from(opcode_cost),
+            },
+            max_gas_burnt: max_gas_burnt,
+            promises_gas: 0,
             prepaid_gas,
             is_view,
             profile: Default::default(),
         }
     }
 
     fn deduct_gas(&mut self, burn_gas: Gas, use_gas: Gas) -> Result<()> {
+        use std::cmp::min;
         assert!(burn_gas <= use_gas);
+        let promise_gas = use_gas - burn_gas;
+        let new_promises_gas =
+            self.promises_gas.checked_add(promise_gas).ok_or(HostError::IntegerOverflow)?;
         let new_burnt_gas =
-            self.burnt_gas.checked_add(burn_gas).ok_or(HostError::IntegerOverflow)?;
-        let new_used_gas = self.used_gas.checked_add(use_gas).ok_or(HostError::IntegerOverflow)?;
+            self.fast_counter.burnt_gas.checked_add(burn_gas).ok_or(HostError::IntegerOverflow)?;
+        let new_used_gas =
+            new_burnt_gas.checked_add(new_promises_gas).ok_or(HostError::IntegerOverflow)?;
         if new_burnt_gas <= self.max_gas_burnt && (self.is_view || new_used_gas <= self.prepaid_gas)
         {
-            self.burnt_gas = new_burnt_gas;
-            self.used_gas = new_used_gas;
+            if promise_gas != 0 {
+                self.fast_counter.gas_limit =
+                    min(self.max_gas_burnt, self.prepaid_gas - new_promises_gas);
+            }
+            self.fast_counter.burnt_gas = new_burnt_gas;
+            self.promises_gas = new_promises_gas;
             Ok(())
         } else {
-            use std::cmp::min;
-            let res = if new_burnt_gas > self.max_gas_burnt {
+            // if max_gas_burnt == prepaid_gas we must use GasExceeded error.
+            if new_burnt_gas > self.max_gas_burnt && self.max_gas_burnt != self.prepaid_gas {
+                self.fast_counter.burnt_gas = self.max_gas_burnt;
                 Err(HostError::GasLimitExceeded.into())
-            } else if new_used_gas > self.prepaid_gas {
-                Err(HostError::GasExceeded.into())
             } else {
-                unreachable!()
-            };
-
-            let max_burnt_gas = min(self.max_gas_burnt, self.prepaid_gas);
-            self.burnt_gas = min(new_burnt_gas, max_burnt_gas);
-            self.used_gas = min(new_used_gas, self.prepaid_gas);
+                self.fast_counter.burnt_gas = min(new_burnt_gas, self.prepaid_gas);
+                // Technically we shall do `self.promises_gas = 0;` or error paths, as in this case
+                // no promises will be kept, but that would mean protocol change.
+                // TODO: consider making this change!
```

Comment:
@olonho could you create an issue for this?

Reply:
Done